### ISSUE TITLE
EMSG events in Google DAI

### ISSUE THREAD

**Issue by avelad** [google/shaka-player#3311]
**Have you read the [Tutorials](http://shaka-player-demo.appspot.com/docs/api/tutorial-welcome.html)?** Yes

**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?** Yes

**What version of Shaka Player are you using?** 3.0.10

**Please ask your question**
I have seen that in the Google DAI project they have uploaded an example of integration with the Shaka Player, in the example they use EMSG events to notify the stream manager. Should the same be done in the internal integration of the Shaka Player?
https://github.com/googleads/googleads-ima-html5-dai/blob/master/shaka/dai.js#L410-L412

I have also seen that the timelineregionenter (example) event is used instead of timelineregionadded (library), what is correct?

@ismena , I think this is of your interest :)

**Comment by ismena** [comment]
Ahh! How come I know so little about the whole emsg event business?! 
If it's working then I probably either wrote or reviewed it, but my mind id just going blank!

So, emsg events signal current ad state, is that right?

**Comment by avelad** [comment]
Well, precisely this part I did not know, that's why I asked you ...

Do you know someone within Google to ask about this?

**Comment by ismena** [comment]
Ah, ok. Sorry, I didn't look at the IMA sample code closely when I wrote the first comment.

I need to sync with the IMA team on that - that example is based on our code from before our IMA integration. If you're using our integration, it shouldn't be relevant for you.

**Tl;Dr:** We're doing the same thing they are and this is already part of the library.

Regarding the way it currently works with library (I did write that code, thanks @TheModMaker for refreshing my memory!)
[This is the part](https://github.com/google/shaka-player/blob/2f65b97b79c9b9dfde1b95afadf7c790827054c7/lib/ads/ad_manager.js#L592) where we handle ad-related EMSG events. There are two ways to communicate timed metadata to the IMA SDK: [processMetaData](https://developers.google.com/interactive-media-ads/docs/sdks/html5/dai/reference/js/StreamManager#processMetadata) and [onTimedMetaData](https://developers.google.com/interactive-media-ads/docs/sdks/html5/dai/reference/js/StreamManager#onTimedMetadata). We use the first one and pass the region start time through it, thus delegating it to the IMA to track when the region was entered. In their example, they're likely using the second one which signals that a region is being entered _now_. Hence the discrepancy between the events we listen to, but 
either way is fine.

Does this make sense?

**Comment by avelad** [comment]
It makes sense to me, but a double check with the IMA team wouldn't be bad.

**Comment by ismena** [comment]
The part about EMSG events is in sync with the IMA team.
I'll reach out to them about using our updated code in their examples.

**Comment by avelad** [comment]
@ismena I've been thinking about this. How would DAI work with HLS using mp4? Right now we process events for DASH, src = (HLS Apple devices), and HLS using mux.js, but there is no ID3 parsing in HLS using mp4, how should it work?

Can you reopen the issue while my question is resolved?

**Comment by ismena** [comment]
Looks like it's [possible](https://www.webcitation.org/6ACo4Nqy6?url=http://www.mp4ra.org/specs.html#id3v2) to embed ID3 in mp4, although we don't parse/use mux to parse them at the moment.

We'll need to do it to support DAI with HLS/mp4. Let me file a separate issue for that. We might have to hunt down some content for testing, though as the IMA team doesn't seem to have any at the moment.

Re: DASH events - the IMA team is working on fixing their samples to reflect our built-in IMA integration.

**Comment by ismena** [comment]
Created #3351

**Comment by avelad** [comment]
Thanks @ismena !
